### PR TITLE
Move `loadbalancerset_status` controller functionality into the `loadbalancer` controller

### PR DESCRIPTION
**Pull request by dergeberl**
Sometimes it happens that the `readyReplicas` and `replicas` in `loadbalancer.status` are not correct.

This PR removes the `loadbalancerset_status` controller and integrates the functionality into the `loadbalancer` controller. Therfore the `loadbalancer` controller now also waches the `LoadBalancerSet` object.

This PR also includes general improvents in the `reconcileLoadBalancerSet` function, to dont need that much requeues.

CC: @timebertt

### CODE REVIEW

File: controllers/yawol-controller/loadbalancer/loadbalancer_controller_test.go [hunk 1]
```diff
@@ -388,6 +388,88 @@ var _ = Describe("loadbalancer controller", Serial, Ordered, func() {
...
+				return nil
+			}, timeout, interval).Should(Succeed())
+
+			By("Wait until replicas in lb object are 0")
+			var actual yawolv1beta1.LoadBalancer
+			Eventually(func() *int {
+				err := k8sClient.Get(ctx, lbNN, &actual)
+				if err != nil {
+					return nil
+				}
```
Comment (timebertt) [inline controllers/yawol-controller/loadbalancer/loadbalancer_controller_test.go:528]: Use a `g.Expect` with `To(Succeed())` here

File: controllers/yawol-controller/loadbalancer/loadbalancer_controller_test.go [hunk 2]
```diff
@@ -388,6 +388,88 @@ var _ = Describe("loadbalancer controller", Serial, Ordered, func() {
...
+				PortName:               pointer.String("port-name"),
+				SecurityGroupID:        pointer.String("sec-group-id"),
+				SecurityGroupName:      pointer.String("sec-group-name"),
+				LastOpenstackReconcile: &timeNow,
+				OpenstackReconcileHash: &reconcileHash,
+			}
+		})
+
+		It("should update the loadbalancer status", func() {
+			lbSet := yawolv1beta1.LoadBalancerSet{}
```
Comment (timebertt) [inline controllers/yawol-controller/loadbalancer/loadbalancer_controller_test.go:512]:
```suggestion
			var lbSet *yawolv1beta1.LoadBalancerSet
```

File: controllers/yawol-controller/loadbalancer/loadbalancer_controller_test.go [hunk 3]
```diff
@@ -388,6 +388,88 @@ var _ = Describe("loadbalancer controller", Serial, Ordered, func() {
...
+			var actual yawolv1beta1.LoadBalancer
+			Eventually(func() *int {
+				err := k8sClient.Get(ctx, lbNN, &actual)
+				if err != nil {
+					return nil
+				}
+				return actual.Status.Replicas
+			}, timeout, interval).Should(Equal(pointer.Int(0)))
+
+			Eventually(func() *int {
```
Comment (timebertt) [inline controllers/yawol-controller/loadbalancer/loadbalancer_controller_test.go:532]: Why is this `Eventually` repeated?

File: controllers/yawol-controller/loadbalancer/loadbalancer_controller_test.go [hunk 4]
```diff
@@ -388,6 +388,88 @@ var _ = Describe("loadbalancer controller", Serial, Ordered, func() {
...
+			Expect(k8sClient.Status().Patch(ctx, &lbSet, patch)).Should(Succeed())
+
+			By("Validate - ExternalIP should be set")
+			Eventually(func() *string {
+				err := k8sClient.Get(ctx, lbNN, &actual)
+				if err != nil {
+					return nil
+				}
+				return actual.Status.ExternalIP
+			}, timeout, interval).ShouldNot(BeNil())
```
Comment (timebertt) [inline controllers/yawol-controller/loadbalancer/loadbalancer_controller_test.go:553]: Why are we checking that `ExternalIP` is set in this test case? If I got it correctly, we set it ourselves earlier.

File: controllers/yawol-controller/loadbalancer/loadbalancer_controller_test.go [hunk 5]
```diff
@@ -388,6 +388,88 @@ var _ = Describe("loadbalancer controller", Serial, Ordered, func() {
...
+				return actual.Status.Replicas
+			}, timeout, interval).Should(Equal(pointer.Int(1)))
+
+			Eventually(func() *int {
+				err := k8sClient.Get(ctx, lbNN, &actual)
+				if err != nil {
+					return nil
+				}
+				return actual.Status.ReadyReplicas
+			}, timeout, interval).Should(Equal(pointer.Int(1)))
```
Comment (timebertt) [inline controllers/yawol-controller/loadbalancer/loadbalancer_controller_test.go:570]: Can you combine both `Eventually`s and perform both assertions with `g.Expect`?

File: internal/helper/loadbalancer.go [hunk 6]
```diff
@@ -164,8 +164,8 @@ func ReadCurrentRevisionFromLB(lb *yawolv1beta1.LoadBalancer) (int, error) {
 
 func GetNextRevisionForLoadBalancer(setList *yawolv1beta1.LoadBalancerSetList) (int, error) {
 	var highestRevision int
-	for _, set := range setList.Items {
-		rev, err := ReadRevisionFromLBS(&set)
+	for i := range setList.Items {
+		rev, err := ReadRevisionFromLBS(&setList.Items[i])
```
Comment (timebertt) [inline internal/helper/loadbalancer.go:168]: Can we keep using the loop variable but disable that stupid linter?

File: controllers/yawol-controller/loadbalancer/loadbalancer_controller_test.go [hunk 7]
```diff
@@ -38,6 +39,106 @@ const (
...
+					Reason:             "Reason",
+					Message:            "Message",
+				},
+			},
+			ReadyReplicas: pointer.Int(1),
+			Replicas:      pointer.Int(1),
+		},
+	}
+
+	It("should reconcile", func() {
```
Comment (timebertt) [inline controllers/yawol-controller/loadbalancer/loadbalancer_controller_test.go:76]: `It`s should be concise and test one specific case. Can you structure it similar to what is done here: ?
https://github.com/gardener/gardener/blob/d48762a67451bb84b8fd21314ef2051eb815880e/pkg/resourcemanager/controller/tokenrequestor/add_test.go

**Review by timebertt — COMMENTED** [review]
Thanks for the PR, I have added some nit suggestions inline.

File: controllers/yawol-controller/loadbalancer/loadbalancer_controller_test.go (hunk 3 above)
Comment (dergeberl) [reply controllers/yawol-controller/loadbalancer/loadbalancer_controller_test.go:532]: One was `replicas`the other `readyReplicas`. I changed it and check both at once.

File: controllers/yawol-controller/loadbalancer/loadbalancer_controller_test.go (hunk 4 above)
Comment (dergeberl) [reply controllers/yawol-controller/loadbalancer/loadbalancer_controller_test.go:553]: 👀 good question, it was done in the old tests of the status controller, but it makes no sense.

File: internal/helper/loadbalancer.go (hunk 6 above)
Comment (dergeberl) [reply internal/helper/loadbalancer.go:168]: We could, but I think it is not that hard to read. There are 2 linters that did not like this code: 
```
               +lint *failed* | internal/helper/loadbalancer.go:167:2: rangeValCopy: each iteration copies 664 bytes (consider pointers or indexing) (gocritic)
               +lint *failed* |         for _, set := range setList.Items {
               +lint *failed* |         ^
               +lint *failed* | internal/helper/loadbalancer.go:168:35: G601: Implicit memory aliasing in for loop. (gosec)
               +lint *failed* |                 rev, err := ReadRevisionFromLBS(&set)
               +lint *failed* |                                                 ^

```

File: controllers/yawol-controller/loadbalancer/loadbalancer_controller_test.go [hunk 8]
```diff
@@ -509,65 +505,37 @@ var _ = Describe("loadbalancer controller", Serial, Ordered, func() {
...
-
-			Eventually(func() *int {
+			Eventually(func(g Gomega) {
 				err := k8sClient.Get(ctx, lbNN, &actual)
-				if err != nil {
-					return nil
-				}
-				return actual.Status.ReadyReplicas
-			}, timeout, interval).Should(Equal(pointer.Int(0)))
+				g.Expect(err).To(Succeed())
```
Comment (timebertt) [inline controllers/yawol-controller/loadbalancer/loadbalancer_controller_test.go]: You could inline the `Get` call into `Expect()`

**Review by breuerfelix — APPROVED** [review]

**Pull request merged**